udp.go

```diff
@@ -0,0 +1,201 @@
+package tunnel
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/nknorg/nkn-sdk-go"
+	ts "github.com/nknorg/nkn-tuna-session"
+	"github.com/nknorg/tuna"
+)
+
+var (
+	ErrUDPWithTuna = errors.New("UDP is only supported in tuna mode")
+)
+
+// Generic interface for UDP conneciton, compatabile to net.UDPConn, nkn-tuna-session.
+type udpConn interface {
+	ReadFrom(b []byte) (n int, addr net.Addr, err error)
+	WriteTo(b []byte, addr net.Addr) (n int, err error)
+	Close() error
+}
+
+// UdpConn with updated (read or read) time.
+// Because UDP is connection-less, we don't know the peer endpoint is alive only when we get data from it.
+// So we purge idle UDP which has no read or write data in configured duration.
+type udpConnWithTime struct {
+	conn udpConn
+
+	sync.RWMutex
+	updTime time.Time // last updated time
+}
+
+func (ucwt *udpConnWithTime) update() {
+	ucwt.Lock()
+	defer ucwt.Unlock()
+	ucwt.updTime = time.Now()
+}
+
+func (ucwt *udpConnWithTime) isOverIdleTime(idleTime int64) bool {
+	ucwt.RLock()
+	defer ucwt.RUnlock()
+	return time.Since(ucwt.updTime) > time.Duration(idleTime*int64(time.Second))
+}
+
+type multiClientDialer struct {
+	*nkn.MultiClient
+}
+
+func newMultiClientDialer(client *nkn.MultiClient) *multiClientDialer {
+	return &multiClientDialer{client}
+}
+
+func (m *multiClientDialer) DialUDPWithConfig(remoteAddr string, config *nkn.DialConfig) (*ts.UdpSession, error) {
+	return nil, ErrUDPWithTuna
+}
+
+func (t *Tunnel) dialUDP(addr string) (udpConn, error) {
+	if t.toNKN {
+		udpSess, err := t.dialer.DialUDPWithConfig(addr, t.config.DialConfig)
+		if err != nil {
+			return nil, err
+		}
+		return udpSess, nil
+	}
+
+	a, err := net.ResolveUDPAddr("udp", addr)
+	if err != nil {
+		fmt.Println(err)
+		return nil, err
+	}
+	conn, err := net.DialUDP("udp", nil, a)
+	if err != nil {
+		return nil, err
+	}
+
+	return conn, nil
+}
+
+func (t *Tunnel) getToUDPConn(from net.Addr) (*udpConnWithTime, bool, error) {
+	t.udpLock.Lock()
+	defer t.udpLock.Unlock()
+
+	toUDPConn, ok := t.udpConns[from.String()]
+	if ok {
+		return toUDPConn, false, nil
+	}
+
+	udpConn, err := t.dialUDP(t.to)
+	if err != nil {
+		return nil, false, err
+	}
+
+	toUDPConn = &udpConnWithTime{conn: udpConn, updTime: time.Now()}
+	t.udpConns[from.String()] = toUDPConn
+
+	return toUDPConn, true, nil
+}
+
+func (t *Tunnel) udpPipe() error {

```

Comment:
This function contains too much stuff that doesn't match its name. I would suggest the following changes:
- Move L104-124 to a new function `listenUDP()`
- Remove L126-128
- Move the content inside goroutine outside

So the caller side will be something like:
```
t.listenUDP()
go t.pipeUDP()
go t.purgeIdleUdpConn()
```

Reply:
Yes, modified.